packages/node-resolve/README.md

```diff
@@ -34,16 +34,23 @@ export default {
   input: 'src/index.js',
   output: {
     dir: 'output',
-    format: 'cjs'
+    format: 'cjs',
   },
-  plugins: [nodeResolve()]
+  plugins: [nodeResolve()],
 };
 ```
 
 Then call `rollup` either via the [CLI](https://www.rollupjs.org/guide/en/#command-line-reference) or the [API](https://www.rollupjs.org/guide/en/#javascript-api).
 
 ## Options
 
+### `exportConditions`
+
+Type: `Array[...String]`<br>
+Default: `['module', 'import']`
```

Comment:
Where is `"module"` defined here? I don't know exactly what its definition is or where that is supported.

Reply:
This is the standard field webpack and rollup will be using.

Reply:
I understand it is a useful way to enable CommonJS modules to load ES modules to avoid duplication cases (although these pains are yet to be felt as there dont seem to be that many people shipping dual mode packages that hit these cases yet).

But I'm not sold that it is necessarily the best way to do this, not that I can't be :)

I just don't think RollupJS should ship this without the meaning being very clearly defined for the various use cases.

Let me just try to get the definition straight then at least:

1. Does `"module"` imply modern syntax, or is that not part of its definition?
2. Does `"module"` imply resolution in a bundler like Webpack or Rollup only? Or is it expected that the definition extends to other bundlers or tools?
3. Does `"module"` apply for the Node.js platform target? This can be seen then as something that changes the default Node.js semantics under bundling which breaks the overall goal of RollupJS building for Node as retaining semantics.
4. Does `"module"` apply for any other platforms other than the browser?
5. Finally, do you think Node.js itself should implement the `"module"` field when resolving `import` (and then just not when resolving `require`)? This seems like it might avoid the repetition of defining `"exports": { "import": "./main.mjs", "module": "./main.mjs", "require": "./main.cjs" }` current redundancy for these sorts of packages.

Reply:
`import` can point to es module which wrap commonjs modules. This is a problem for pure es module workflows, so there needs to be a way to indicate pure es modules. The name `module` came out of the discussion at https://github.com/rollup/plugins/issues/362 and https://github.com/webpack/webpack/issues/11014#issuecomment-643256943

I'm fine with deferring that to a later PR, but it would be a breaking change later on again. @lukastaegert what do you think?

Reply:
Thanks, I'm not against it, I just want to be clear on the meaning. Ok, it seems the key source here is Tobias's proposal. Reading his proposal more clearly, let me try and answer my own questions then:

> Does "module" imply modern syntax, or is that not part of its definition?

Not mentioned in the doc

> Does "module" imply resolution in a bundler like Webpack or Rollup only? Or is it expected that the definition extends to other bundlers or tools?

Not mentioned in the doc.

> Does "module" apply for the Node.js platform target? This can be seen then as something that changes the default Node.js semantics under bundling which breaks the overall goal of RollupJS building for Node as retaining semantics.

The doc explicitly mentions that you would want to use it when bundling for a Node.js target so it doesn't seem to be Node or browser specific at all as far as I can tell.

> Does "module" apply for any other platforms other than the browser?

As above, it can apply to Node.js

> Finally, do you think Node.js itself should implement the "module" field when resolving import (and then just not when resolving require)? This seems like it might avoid the repetition of defining "exports": { "import": "./main.mjs", "module": "./main.mjs", "require": "./main.cjs" } current redundancy for these sorts of packages.

Node.js likely would not do this.

Basically it seems that "module" is specifically there to solve the bundler problem for when CJS is able to import from ESM in bundlers only. And RollupJS as a bundler can join Webpack in doing that.

These constraints seem clear to me, I hope we can give users a little guidance on when to use "module" mainly.

@lukastaegert will entirely defer to you on this one, I'm happy either way, just erring on the side of caution in introducing new things.

Reply:
As you correctly stated, this condition was specifically created for the bundler use case and I definitely think we should add it by default. An entirely different question is when library authors should use it. Here, I would see it as a last resort solution to solve e.g. certain dual state hazard situations in a purely ESM way, so there are other patterns I would recommend first.